pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -569,15 +569,18 @@ func (cb *ClusterBuilder) buildLocalityLbEndpoints(proxyView model.ProxyView, se
 			},
 			Metadata: &core.Metadata{},
 		}
+		var metadata *model.EndpointMetadata
 
-		metadata := instance.Endpoint.Metadata()
 		if features.CanonicalServiceForMeshExternalServiceEntry && service.MeshExternal {
-			metadata.Namespace = service.Attributes.Namespace
 			svcLabels := service.Attributes.Labels
 			if _, ok := svcLabels[model.IstioCanonicalServiceLabelName]; ok {
+				metadata = instance.Endpoint.MetadataClone()
 				metadata.Labels[model.IstioCanonicalServiceLabelName] = svcLabels[model.IstioCanonicalServiceLabelName]
 				metadata.Labels[model.IstioCanonicalServiceRevisionLabelName] = svcLabels[model.IstioCanonicalServiceRevisionLabelName]
+			} else {
+				metadata = instance.Endpoint.Metadata()
 			}
+			metadata.Namespace = service.Attributes.Namespace
 		}
 		util.AppendLbEndpointMetadata(metadata, ep.Metadata)
```

Comment:
I think `metadata` is unset now if we don't match L574?

Reply:
fixed